### PR TITLE
crypto: use fastcrypto Hex everywhere

### PR DESCRIPTION
**Pull request by joyqvq**

### CODE REVIEW

**Review by jonas-lj — APPROVED** [review]
Nice!

File: crates/sui-types/src/base_types.rs [hunk 1]
```diff
@@ -474,8 +474,8 @@ pub fn context_from_digest(digest: TransactionDigest) -> Context {
     // TODO: don't create a HashMap, that wastes memory and costs an allocation!
     let mut carrier = HashMap::new();
     // TODO: figure out exactly what key to use.  I suspect it has to be the parent span ID in OpenTelemetry format.
-    carrier.insert("traceparent".to_string(), hex::encode(digest.0));
-    // carrier.insert("tx_digest".to_string(), hex::encode(digest.0));
+    carrier.insert("traceparent".to_string(), Hex::encode(digest.0));
```
Comment (kchalkias) [inline crates/sui-types/src/base_types.rs:477]: These do not have the same behavior as previously. 

Honestly, I got very confused when I realized that somehow in fastcrypto we switched the behavior of 

- previously `hex:encode` didn't attach `0x`, while fastcrypto's `Hex::encode` attaches `0x` 
- previously `encode_bytes_hex` attached `0x`, while fastcrypto's `encode_bytes_hex` doesn't.

IMO the previous behavior was better. It's probably a mistake that our fastcrypto's `Hex::encode` attaches `0x` (this is important for compatibility with other Hex libs + note that the 0x-format is optional for a generic app that uses fastcrypto, and it just happens to be a helper for us).

TL;DR I'd expect the helper function `encode_bytes_hex` to attach the `0x`. We also need better name, cause it's very difficult to spot the difference between the two unctions while reviewing a crate that uses fastcrypto.

My suggestion: fix fastcrypto first, then update Sui, wdyt? @joyqvq @jonas-lj ?

**Review by kchalkias — CHANGES_REQUESTED** [review]

File: crates/sui-types/src/crypto.rs [hunk 2]
```diff
@@ -352,7 +352,7 @@ pub struct ConciseAuthorityPublicKeyBytes<'a>(&'a AuthorityPublicKeyBytes);
 
 impl std::fmt::Debug for ConciseAuthorityPublicKeyBytes<'_> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
-        let s = hex::encode(&self.0 .0[0..4]);
+        let s = encode_bytes_hex(&self.0 .0[0..4]);
```
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:355]: Taking the opportunity here. Per our recent SEV re panicking when accessing short slices or vectors, I'd personally add an `assert` to check the payload has enough bytes (at least 4) before doing`self.0[0..4]`. 

In particular for public keys, we know that current implementation is probably safe, but imagine what if @jonas-lj one day builds a naive Sig scheme where its pubKey is super short ie just 2-3 bytes.

That said, I'm opening an issue to check all subslice subvectors for panicking in all Sui's repos, starting from Sui, fastcrypto and MystenInfra. Lmk who wants to work on that.

**Review by kchalkias — CHANGES_REQUESTED** [review]

File: crates/sui-json/src/tests.rs [hunk 3]
```diff
@@ -334,7 +335,7 @@ fn test_basic_args_linter_pure_args_good() {
         (
             Value::from(good_hex_val),
             MoveTypeLayout::Vector(Box::new(MoveTypeLayout::U8)),
-            bcs::to_bytes(&hex::decode(good_hex_val.trim_start_matches(HEX_PREFIX)).unwrap())
+            bcs::to_bytes(&Hex::decode(good_hex_val.trim_start_matches(HEX_PREFIX)).unwrap())
```
Comment (kchalkias) [inline crates/sui-json/src/tests.rs:338]: which of the two functions do we need here?

File: crates/sui-types/src/base_types.rs (hunk 1 above)
Comment (joyqvq) [reply crates/sui-types/src/base_types.rs:477]: https://github.com/MystenLabs/fastcrypto/blob/main/fastcrypto/src/encoding.rs#L155

to be clear `encode_bytes_hex` had always not added 0x and its just `hex::encode`. `Hex::encode` does add 0x. 

there are places we do want 0x and places we dont in Sui, in this PR we follow:

1. when we dont want 0x -> encode_bytes_hex
2. when we want 0x -> Hex::encode

File: crates/sui-types/src/crypto.rs (hunk 2 above)
Comment (joyqvq) [reply crates/sui-types/src/crypto.rs:355]: good idea!

File: crates/sui-json/src/tests.rs (hunk 3 above)
Comment (joyqvq) [reply crates/sui-json/src/tests.rs:338]: so Hex::decode tries to strip 0x and then hex::decode. effectively the same

File: crates/sui-types/src/base_types.rs [hunk 4]
```diff
@@ -546,16 +546,6 @@ where
     Ok(value)
 }
 
-pub fn encode_bytes_hex<B: AsRef<[u8]>>(bytes: B) -> String {
-    hex::encode(bytes.as_ref())
```
[the remark this answers is not in the document]
Comment (joyqvq) [reply crates/sui-types/src/base_types.rs:550]: so encode_bytes_hex is also from fastcrypto (this function is purely moved inside fastcrypto but not replaced by Hex::encode)

File: crates/sui-types/src/base_types.rs (hunk 1 above)
Comment (kchalkias) [reply crates/sui-types/src/base_types.rs:477]: For consistency w other hex libs out there and because the 0x is just a fronting convention I'd prefer the struct's main methods to not deal w formats, but only the helper functions should do

File: Cargo.toml [hunk 5]
```diff
@@ -115,8 +115,8 @@ move-ir-types = { git = "https://github.com/move-language/move", rev = "0800fc79
 move-prover = { git = "https://github.com/move-language/move", rev = "0800fc79a98ca304bd449878f545cdcaff6f94bb" }
 move-prover-boogie-backend = { git = "https://github.com/move-language/move", rev = "0800fc79a98ca304bd449878f545cdcaff6f94bb" }
 
-fastcrypto = { git = "https://github.com/MystenLabs/fastcrypto", rev = "12c522bea01e4a3d19ecae63d789edcdcce2d66c" }
-fastcrypto-zkp = { git = "https://github.com/MystenLabs/fastcrypto", rev = "12c522bea01e4a3d19ecae63d789edcdcce2d66c", package = "fastcrypto-zkp" }
+fastcrypto = { git = "https://github.com/MystenLabs/fastcrypto", rev = "b6411e014255c9e2a0f6dff00478bc45d76e9f5d" }
```
Comment (kchalkias) [inline Cargo.toml:118]: It is ok to have different dep versions of fastcrypto across our Sui codebase? Or when we change in one place we should change everywhere?

**Review by kchalkias — APPROVED** [review]

**Review by huitseeker — COMMENTED** [review]
Would it be possible to take this as an occasion to make fastcrypto a workspace dependency?

File: Cargo.toml (hunk 5 above)
Comment (joyqvq) [reply Cargo.toml:118]: b6411e is the only version used in Sui. all crates refer to this workspace `fastcrypto = { workspace = true }` cargo to derive version

**Comment by joyqvq** [comment]
> Would it be possible to take this as an occasion to make fastcrypto a workspace dependency?

this is already done! https://github.com/MystenLabs/sui/pull/5115/files

**Review by kchalkias — APPROVED** [review]
